core/src/main/java/io/parsingdata/metal/token/TokenRef.java

```diff
@@ -60,22 +64,20 @@ public TokenRef(final String name, final String referenceName, final Encoding en
 
     @Override
     protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
-        return lookup(environment.order, referenceName).parse(scope, environment, encoding);
+        return lookup(ImmutableList.create(environment.order), referenceName).computeResult().parse(scope, environment, encoding);
     }
 
-    private Token lookup(final ParseItem item, final String referenceName) {
-        if (item.getDefinition().name.equals(referenceName)) {
-            return item.getDefinition();
-        }
-        if (!item.isGraph() || item.asGraph().isEmpty()) { return LOOKUP_FAILED; }
-        final Token headResult = lookup(item.asGraph().head, referenceName);
-        if (headResult != LOOKUP_FAILED) { return headResult; }
-        return lookup(item.asGraph().tail, referenceName);
+    private SafeTrampoline<Token> lookup(final ImmutableList<ParseItem> items, final String referenceName) {
+        if (items.isEmpty()) { return complete(() -> LOOKUP_FAILED); }
+        final ParseItem item = items.head;
+        if (item.getDefinition().name.equals(referenceName)) { return complete(item::getDefinition); }
+        if (item.isGraph() && !item.asGraph().isEmpty()) { return intermediate(() -> lookup(items.tail.add(item.asGraph().tail).add(item.asGraph().head), referenceName)); }
```

Comment:
The rule was the put the return of a value on the same line, but actual code on a new line. So I agree on the bottom one but the others are just values.